### DISCUSSION TITLE
[Proposal] Add exit section that will catch any result in the function

### DISCUSSION THREAD

**Discussion by ichensky** [dotnet/csharplang#5644] [Language Ideas]
The proposal is to add `exit` section that will catch any result in the function even if it is an `Exception`. 
We can put `exit` in any place inside of the function. 
This section will allow adding some cross-cutting functionality like logging, caching, or post preparation and validating results.

```csharp
        public Dictionary<int, int> cache = new();

        public int Foo(int id, string name, int status)
        {
            if (cache.TryGetValue(id, out int result))
                return result;
            
            if (id != 100500)
                return -1;

            if (status > 100)
                throw new ArgumentException();

            if (name != "Bob" && status != 2)
            {
                // do some logic ..
                return 10;
            }

            if (status)
            {
                // do some logic ..
            }
            
            // do some logic ..

            return 0;


            exit (Result<int> result)
            {
                if (result.Exception != null)
                {
                    logger.LogError(result.Exception, $"... id:{id}; status: {status}");

                    // just return from exit and `Foo`
                    return;
                }
                else
                {
                    logger.LogDebug($"... original result {result.Value}");

                    if (result.Value == -1)
                        result.Value = 0;

                    cache[id] = result.Value;

                    logger.LogInformation($"... {result.Value}");
                }
            }
        }
```

**Comment by Xyncgas** [comment]
the exit section you are talking about is the run time, where as inside your code you can put a big try {} catch{} inside your entry point / event loop

**Reply by ichensky** [reply]
`try {} catch{}` catch only an exception, but the idea is to catch function result. Of course, it is possible to wrap one function into another, but sometimes it's better to keep some functionality in one block of code, that is why `exit` section could be useful.

**Reply by Xyncgas** [reply]
I admit I was narrow minded

**Reply by Xyncgas** [reply]
I mean how do you think exceptions work, isn't exception implemented just like your exit section in assembly already or I am shooting at the wrong direction?

**Reply by Xyncgas** [reply]
what I mean is that, whenever something unwanted happen, control is yielded that's exception mechanism, you didn't have to throw exceptions ok imagine this world is without exceptions then we expect things still work until it breaks, if however you would rather see the exceptions there and deal with it, it works just like how your exit section does by executing when something is wrong, and this propagate all the way to the top which is obviously something the exit section is trying to prevent, so I would suggest something beautiful in addition to yours :Have a synctax sugar like ```catch { }``` without ```try``` in the body and you can do whatever you want there logging etc

**Comment by theunrepentantgeek** [comment]
This behaviour can already be achieved fairly simply with idiomatic C# - here's one way to do that, rewriting the original sample code (please forgive any errors):

``` csharp
public int Foo(int id, string name, int status)
{
	try
	{
	    var result = FooImpl()
	}
	catch (Exception ex)
	{
		logger.LogError(ex, $"... id:{id}; status: {status}");
	}
	finally
	{
		logger.LogDebug($"... original result {result}");

		if (result == -1)
			result = 0;

		cache[id] = result;

		logger.LogInformation($"... {result}");
	}
	
	int FooImpl()
	{
		if (cache.TryGetValue(id, out int result))
			return result;
		
		if (id != 100500)
			return -1;

		if (status > 100)
			throw new ArgumentException();

		if (name != "Bob" && status != 2)
		{
			// do some logic ..
			return 10;
		}

		if (status)
		{
			// do some logic ..
		}
		
		// do some logic ..

		return 0;
	}
}
```

The key question then is this: What's so wrong with this approach that a new keyword, such as **exit**, is warranted? 
How does it go from the traditional [-100 points](https://docs.microsoft.com/en-gb/archive/blogs/ericgu/minus-100-points) to a high enough score to be worth the (decidedly non-trivial) implementation and adoption cost?

**Comment by VardanKosakyan** [comment]
It behaves like interceptors in AOP

**Comment by huoyaoyuan** [comment]
This is exactly a duplication of #1398, which is rejected in favor of `try-finally`.
